### ISSUE TITLE
Evaluate ensemble weighting: inverse-MAPE blend trails best-base on 47/51 BAs

### ISSUE THREAD

**Issue by kristenmartino** [kristenmartino/gridpulse#181]
## Context

With the holdout-NaN crash fixed (#176 / PR #179) the ensemble holdout column is
now populated for all 51 BAs (PR #180 refreshed the docs). The first full-data
look raises a real question about the current ensemble design.

## Observation

From the 2026-06-19 holdout (`scripts/export_holdout_metrics.py`, all 51 BAs):

| Distribution | min | median | mean | p90 | max |
|---|---|---|---|---|---|
| XGBoost-only | 0.98% | 2.32% | 3.79% | 6.57% | 33.97% (AZPS) |
| Best-base per BA | 0.98% | 2.30% | 3.61% | 6.57% | 26.68% (AZPS) |
| **Ensemble (served)** | 1.70% | **3.48%** | **4.92%** | **8.37%** | 27.40% (AZPS) |

- The **inverse-MAPE ensemble (ADR-004) trails best-base in aggregate** —
  median 3.48% vs 2.30%, p90 8.37% vs 6.57%.
- The ensemble beats **XGBoost-alone on only 4 of 51 BAs.**
- XGBoost is the best base model for **48 of 51** BAs (ARIMA wins AZPS + GCPD,
  Prophet wins SPA).

Mechanically this is expected: `weight_i = 1/MAPE_i` (normalized) still assigns
meaningful weight to Prophet and ARIMA even when they're 3–5× worse than
XGBoost, so the blend lands above the strongest single model. The ensemble's
genuine benefit is **tail variance-reduction** (AZPS: XGBoost 33.97% → blend
27.40%), not central-tendency accuracy.

## Question to resolve

Is serving the inverse-MAPE ensemble as the default forecast still the right
call, given it underperforms both XGBoost-alone and best-base-per-BA on the
median BA? Options to evaluate (not prejudging):

1. **Keep as-is.** Accept the median accuracy cost for the tail robustness +
   simplicity; document the trade-off (partly done in PR #180). Cheapest.
2. **Sharpen the weighting.** Use `1/MAPE^k` (k>1) or softmax-over-skill so the
   best model dominates unless a peer is genuinely close — recovers most of the
   XGBoost accuracy while still blending when models are comparable.
3. **Best-base selection as default.** Serve the per-BA best holdout model
   (mostly XGBoost, ARIMA for AZPS/GCPD, Prophet for SPA). Loses ensemble
   variance-reduction; risks per-BA model flapping run-to-run (see below).
4. **Threshold/guardrail.** Only blend models whose holdout MAPE is within X× of
   the best; otherwise serve the best single model.

## Constraints / things to check during evaluation

- **Run-to-run variance is real:** AZPS best-base went 11.90% (2026-06-17) →
  26.68% (2026-06-19). Any best-base/selection scheme must be stable against
  this (e.g. smoothed/rolling holdout) or it'll flap the served model daily.
- Decide the metric of record: this compares **168h holdout MAPE**; the
  live-served accuracy is the 1h-ahead drift metric (`models/drift.py`), which
  may rank models differently. Worth confirming the holdout ranking matches the
  live ranking before re-weighting on holdout numbers.
- Whatever changes, the **Models tab** + `CANONICAL_FACTS.md` /
  `BACKTEST_RESULTS.md` framing must stay honest about which number is
  "served" vs "best achievable."

## Proposed empirical methodology

How to choose weights from data instead of from a formula. The current
`compute_ensemble_weights` (`models/ensemble.py:16`) is a closed-form heuristic
fit to *nothing* — it assumes the optimal blend is inversely proportional to
each model's solo MAPE. "Optimize empirically" = for each BA, pick the weight
vector that minimizes realized error on held-out data.

### Gating prerequisite: persist the per-hour holdout vectors

To fit weights empirically we need the per-hour holdout prediction vectors per
model — `ŷ_xgb[t], ŷ_prophet[t], ŷ_arima[t], y_actual[t]` over the 168h window.
The training job already **computes** exactly these
(`jobs/training_job.py:235`, `payload["forecast"]` / `payload["actual"]`) but
**only the scalar MAPE is persisted** to GCS `meta.json`; the vectors are
discarded after the run. So:

- We **cannot** optimize offline from what's in GCS today (only scalars survive).
- **Step 0 of every empirical method below** is persisting the holdout vectors
  (extend the `extra` payload written by `_train_*`) or a dedicated backtest
  that regenerates them. This is the single blocking infra change — without it
  every method here is blocked. Vectors are small (3 × 168 floats per BA).

### The optimization (per BA, over the holdout window)

Constrained least squares on the simplex:

```
minimize_w   Σ_t ( y_t − Σ_i w_i · ŷ_i,t )²
subject to   Σ w_i = 1,  w_i ≥ 0
```

- Convex QP — `scipy.optimize.nnls` or `minimize(method="SLSQP")` solves it in
  ms for 3 models × 168 points.
- The non-negativity + sum-to-1 constraint keeps weights interpretable and
  preserves the existing "ensemble bounded by member min/max" guarantee
  (`ensemble.py:99`). Unconstrained regression would let weights go
  negative/explode and overfit 168 points.
- Squared error optimizes RMSE; to optimize MAPE (what we report) weight each
  residual by `1/y_t` — still convex.

### Ladder of methods to compare (simplest → most expressive)

1. **Constrained LS per BA** (above) — the honest "optimal blend" reference.
   Biggest expected win, simplest to reason about.
2. **Sharpened inverse-MAPE: `w_i ∝ (1/MAPE_i)^k`** — one global hyperparameter.
   `k=1` is today; `k→∞` is best-base selection; `k≈2–4` usually captures most
   of the LS gain. Tune `k` once across all 51 BAs. Cheapest change, **no
   per-BA state to persist or drift**, low overfitting risk → recommended first
   move if a change is warranted.
3. **Best-base selection** (`k→∞` limit) — serve `argmin MAPE` per BA. Data
   already says this beats the current ensemble on central accuracy; cost is
   losing the tail variance-reduction the blend buys.
4. **Threshold guardrail** — drop any base whose holdout MAPE exceeds X× the
   best, then inverse-MAPE blend the survivors. Collapses to best-base where the
   weak models are hopeless, keeps a blend where bases are competitive.

### Validation methodology (the leakage trap)

Fitting `w` to the same 168h then *reporting* on it is leakage — a great
in-sample number that won't hold live. The comparison **must** be
walk-forward / nested:

- Fit `w` on `[t−336, t−168]`, evaluate on `[t−168, t]`. Roll across several
  windows.
- Compare each strategy (current 1/MAPE, LS, sharpened-`k`, best-base,
  threshold) on the *out-of-fit* windows, per BA. Report a **distribution
  across windows**, not a single snapshot.
- Stability is a first-class criterion: a method that wins on one snapshot but
  flaps across windows (cf. AZPS 11.90%→26.68%) is worse than the boring
  incumbent.
- State the objective explicitly — median accuracy points to best-base/high-`k`;
  tail robustness points to keeping a constrained blend. Different objectives
  give different winners.

### Recommended sequence

1. **Persist holdout vectors** (gating infra; small `extra`-payload change in
   `_train_*`). Suggest splitting this into its own small first PR.
2. Build an **offline harness** in `scripts/` that reads the vectors and runs
   the walk-forward comparison → a table mirroring `docs/BACKTEST_RESULTS.md`.
3. Only if a method wins out-of-fit across windows, change
   `compute_ensemble_weights` — keeping the simplex constraint so the bounded
   guarantee holds. Start with **sharpened-`k`** (one global knob, no per-BA
   state).

## Pointers

- Weighting: `models/ensemble.py`
- Holdout computation: `jobs/training_job.py` (`_ensemble_holdout_metrics`,
  `_holdout_metrics_*`)
- Current numbers: `docs/BACKTEST_RESULTS.md` (post-PR #180)
- ADRs: ADR-004 (1/MAPE weighted ensemble), ADR-005 (XGBoost primary)

Refs #176. Surfaced while verifying the #176 fix in prod.

**Comment by kristenmartino** [comment]
**Measured on the recursive re-measure (2026-07-03, all 51 BAs):**
- Current inverse-MAPE ensemble median: **4.82%**
- Best-model-per-BA (lowest-holdout-MAPE base model per BA): **4.12%** — beats the current ensemble on **36/51 BAs**.
- Base-model wins: **XGBoost 44 / Prophet 4 / ARIMA 3**.

So the blend costs ~0.7pp median vs just serving the best model per BA. **Best-model-per-BA is the cleaner fix than the exponent** (`ENSEMBLE_WEIGHT_EXPONENT`, staged in #229): it's deterministic, is the k→∞ ceiling the exponent asymptotes toward, and matches the fact that the headline `predicted_demand_mw` is already XGBoost-primary. Recommend per-BA best-model selection, retaining the blend only where the top-2 base models are within ~1.2×. Validate on the next retrain before flipping production.
